plugins/ui-update/components/UIDetails.tsx

```diff
@@ -0,0 +1,218 @@
+import { BehaviorSubject, combineLatest, merge, Observable, of } from "rxjs";
+import gql from "graphql-tag";
+
+import ConfigurationMapSection from "#SRC/js/components/ConfigurationMapSection";
+import { schema as cosmosSchema } from "#SRC/js/data/cosmos";
+import { Package } from "#SRC/js/data/cosmos/Package";
+import { default as uiServiceSchema } from "#SRC/js/data/ui-update";
+import {
+  DEFAULT_UI_METADATA,
+  UIMetadata
+} from "#SRC/js/data/ui-update/UIMetadata";
+
+import AvailableUpdateRow from "#PLUGINS/ui-update/components/AvailableUpdateRow";
+import InstalledVersion from "#PLUGINS/ui-update/components/InstalledVersion";
+import InitialScreen from "#PLUGINS/ui-update/components/InitialScreen";
+import FallbackScreen from "#PLUGINS/ui-update/components/FallbackScreen";
+
+import {
+  EMPTY_ACTION,
+  UIAction,
+  UIActions,
+  UIActionType
+} from "#PLUGINS/ui-update/types/UIAction";
+import { versionUpdateAvailable } from "#PLUGINS/ui-update/utils";
+
+interface MutationError {
+  message: string;
+  name: string;
+}
+interface UIUpdateResult {
+  data: { updateDCOSUI: string };
+}
+interface UIRollbackResult {
+  data: { resetDCOSUI: string | null };
+}
+
+const uiServiceActions$ = new BehaviorSubject<UIAction>(EMPTY_ACTION);
+
+function queryCosmosForUIVersions(): Observable<Package> {
+  return graphqlObservable<{ package: Package }>(
+    gql`
+      query {
+        package(name: $packageName) {
+          name
+          versions
+        }
+      }
+    `,
+    cosmosSchema,
+    { packageName: "dcos-ui" }
+  ).pipe(map(result => result.data.package));
+}
+
+function queryUIServiceForMetadata(): Observable<UIMetadata> {
+  return graphqlObservable<{ ui: UIMetadata }>(
+    gql`
+      query {
+        ui {
+          clientBuild
+          packageVersion
+          packageVersionIsDefault
+          serverBuild
+        }
+      }
+    `,
+    uiServiceSchema,
+    { packageName: "dcos-ui" }
+  ).pipe(
+    map(result => result.data.ui),
+    catchError(() => of(DEFAULT_UI_METADATA))
+  );
+}
+
+function onClickRollbackUI() {
+  rollbackUI();
+}
+
+function rollbackUI(delayMs: number = 45000) {
+  uiServiceActions$.next({
+    type: UIActionType.Reset,
+    action: UIActions.Started,
+    value: ""
+  });
+  graphqlObservable<{ resetDCOSUI: string | null }>(
+    gql`
+      mutation {
+        resetDCOSUI
+      }
+    `,
+    uiServiceSchema,
+    {}
+  )
+    .pipe(
+      take(1),
+      delay(delayMs)
+    )
+    .subscribe(
+      (result: UIRollbackResult) => {
+        uiServiceActions$.next({
+          type: UIActionType.Reset,
+          action: UIActions.Completed,
+          value: result.data.resetDCOSUI || ""
+        });
+      },
+      (error: MutationError) => {
+        uiServiceActions$.next({
+          type: UIActionType.Reset,
+          action: UIActions.Error,
+          value: error.message
+        });
+      }
+    );
+}
+
+function onClickStartUpdate(version: string, delayMs: number = 45000) {
```

Comment:
I find it a little odd that we export event handlers from the module, but I get why: this is a method you want to test. 

What if you extract this as a query or update command, rename the function to handleClickStartUpdate and just have it invoke the query or command which is tested separately. That could help you break  up the component and simplify the exports

Reply:
I'm only exporting them to be able to test them.

Reply:
yeah that's my point-- seems like if its complex enough to be tested in isolation it could be its own module